OpenRA.Mods.AS/Traits/TeleportNetworkTransportable.cs

```diff
@@ -0,0 +1,140 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2015- OpenRA.Mods.AS Developers (see AUTHORS)
+ * This file is a part of a third-party plugin for OpenRA, which is
+ * free software. It is made available to you under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation. For more information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Mods.AS.Activities;
+using OpenRA.Mods.Common;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Orders;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.AS.Traits
+{
+	[Desc("Can move instantly to primary designated nydus canal actors.")]
+	class TeleportNetworkTransportableInfo : ITraitInfo
+	{
+		[VoiceReference]
+		public readonly string Voice = "Action";
+		public readonly string EnterCursor = "enter";
+		public readonly string EnterBlockedCursor = "enter-blocked";
+		public object Create(ActorInitializer init) { return new TeleportNetworkTransportable(init, this); }
+	}
+
+	class TeleportNetworkTransportable : IIssueOrder, IResolveOrder, IOrderVoice
+	{
+		readonly TeleportNetworkTransportableInfo info;
+
+		public TeleportNetworkTransportable(ActorInitializer init, TeleportNetworkTransportableInfo info)
+		{
+			this.info = info;
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get { yield return new TeleportNetworkTransportOrderTargeter(info); }
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (order.OrderID != "TeleportNetworkTransport")
+				return null;
+
+			// You can't enter enemy nydus canal so this will do.
+			if (target.Type == TargetType.FrozenActor)
+				return null;
+
+			return new Order(order.OrderID, self, target, queued) { };
+		}
+
+		// Does the player own enough teleport network canals?
+		// For teleport network canals to work, the player needs at least 2.
+		// Actually, at least 1 but in that case, the unit goes in and immediately comes out. That's stupid.
+		static bool HasEnoughCanals(Actor owner)
+		{
+			var teleporttrait = owner.Trait<TeleportNetwork>();
```

Comment:
Type lookup missing. Will probably need the type being passed.

Reply:
Oh, this owner is the actor. Then rename the variable, because this is confusing as hell + you need to check for each traits and not stop beyond the first one.

Reply:
This returns to the issue that this indeed needs a type by now though (to know which type you're looking up at the manager). The EnterTeleportNetwork trait should also have a singular type property.